File: data-structures/go/src/linkedlist/linked_list.go
package linkedlist

import "errors"

var (
	ErrOutOfRange = errors.New("LinkedList: index out of range")
	ErrEmptyList  = errors.New("LinkedList: list is empty")
)

type node[T any] struct {
	value T
	next  *node[T]
}

type LinkedList[T any] struct {
	head *node[T]
	tail *node[T]
	size int
}

func New[T any]() *LinkedList[T] {
	return &LinkedList[T]{}
}

func (l *LinkedList[T]) Size() int {
	return l.size
}

func (l *LinkedList[T]) IsEmpty() bool {
	return l.size == 0
}

func (l *LinkedList[T]) PushFront(value T) {
	n := &node[T]{value: value, next: l.head}
	l.head = n
	if l.tail == nil {
		l.tail = n
	}
	l.size++
}

func (l *LinkedList[T]) PushBack(value T) {
	n := &node[T]{value: value}
	if l.tail == nil {
		l.head = n
		l.tail = n
	} else {
		l.tail.next = n
		l.tail = n
	}
	l.size++
}

func (l *LinkedList[T]) PopFront() (T, error) {
	var zero T
	if l.head == nil {
		return zero, ErrEmptyList
	}
	value := l.head.value
	l.head = l.head.next
	if l.head == nil {
		l.tail = nil
	}
	l.size--
	return value, nil
}

func (l *LinkedList[T]) PopBack() (T, error) {
	var zero T
	if l.head == nil {
		return zero, ErrEmptyList
	}
	if l.head == l.tail {
		value := l.head.value
		l.head = nil
		l.tail = nil
		l.size--
		return value, nil
	}
	curr := l.head
	for curr.next != l.tail {
		curr = curr.next
	}
	value := l.tail.value
	curr.next = nil
	l.tail = curr
	l.size--
	return value, nil
}

func (l *LinkedList[T]) Front() (T, error) {
	var zero T
	if l.head == nil {
		return zero, ErrEmptyList
	}
	return l.head.value, nil
}

func (l *LinkedList[T]) Back() (T, error) {
	var zero T
	if l.tail == nil {
		return zero, ErrEmptyList
	}
	return l.tail.value, nil
}

func (l *LinkedList[T]) At(index int) (T, error) {
	var zero T
	if index < 0 || index >= l.size {
		return zero, ErrOutOfRange
	}
	curr := l.head
	for i := 0; i < index; i++ {
		curr = curr.next
	}
	return curr.value, nil
}

func (l *LinkedList[T]) InsertAt(index int, value T) error {
	if index < 0 || index > l.size {
		return ErrOutOfRange
	}
	if index == 0 {
		l.PushFront(value)
		return nil
	}
	if index == l.size {
		l.PushBack(value)
		return nil
	}
	curr := l.head
	for i := 0; i < index-1; i++ {
		curr = curr.next
	}
	n := &node[T]{value: value, next: curr.next}
	curr.next = n
	l.size++
	return nil
}

func (l *LinkedList[T]) RemoveAt(index int) (T, error) {
	var zero T
	if index < 0 || index >= l.size {
		return zero, ErrOutOfRange
	}
	if index == 0 {
		return l.PopFront()
	}
	curr := l.head
	for i := 0; i < index-1; i++ {
		curr = curr.next
	}
	value := curr.next.value
	curr.next = curr.next.next
	if curr.next == nil {
		l.tail = curr
	}
	l.size--
	return value, nil
}

func (l *LinkedList[T]) Clear() {
	l.head = nil
	l.tail = nil
	l.size = 0
}

func (l *LinkedList[T]) Clone() *LinkedList[T] {
	clone := New[T]()
	curr := l.head
	for curr != nil {
		clone.PushBack(curr.value)
		curr = curr.next
	}
	return clone
}

func (l *LinkedList[T]) Values() []T {
	if l.size == 0 {
		return nil
	}
	result := make([]T, l.size)
	curr := l.head
	for i := 0; curr != nil; i++ {
		result[i] = curr.value
		curr = curr.next
	}
	return result
}

File: data-structures/python/src/linked_list.py
class LinkedList:
    class Node:
        def __init__(self, value):
            self.value = value
            self.next = None

    def __init__(self):
        self._head = None
        self._tail = None
        self._size = 0

    def push_front(self, value):
        node = self.Node(value)
        node.next = self._head
        self._head = node
        if self._tail is None:
            self._tail = node
        self._size += 1

    def push_back(self, value):
        node = self.Node(value)
        if self._tail is None:
            self._head = node
            self._tail = node
        else:
            self._tail.next = node
            self._tail = node
        self._size += 1

    def pop_front(self):
        if self._head is None:
            raise IndexError("pop_front from empty list")
        value = self._head.value
        self._head = self._head.next
        if self._head is None:
            self._tail = None
        self._size -= 1
        return value

    def pop_back(self):
        if self._head is None:
            raise IndexError("pop_back from empty list")
        value = self._tail.value
        if self._head is self._tail:
            self._head = None
            self._tail = None
        else:
            current = self._head
            while current.next is not self._tail:
                current = current.next
            current.next = None
            self._tail = current
        self._size -= 1
        return value

    def front(self):
        if self._head is None:
            raise IndexError("front from empty list")
        return self._head.value

    def back(self):
        if self._tail is None:
            raise IndexError("back from empty list")
        return self._tail.value

    def at(self, index):
        if not isinstance(index, int):
            raise TypeError("index must be an integer")
        if index < 0 or index >= self._size:
            raise IndexError("index out of range")
        current = self._head
        for _ in range(index):
            current = current.next
        return current.value

    def insert_at(self, index, value):
        if not isinstance(index, int):
            raise TypeError("index must be an integer")
        if index < 0 or index > self._size:
            raise IndexError("index out of range")
        if index == 0:
            self.push_front(value)
        elif index == self._size:
            self.push_back(value)
        else:
            node = self.Node(value)
            current = self._head
            for _ in range(index - 1):
                current = current.next
            node.next = current.next
            current.next = node
            self._size += 1

    def remove_at(self, index):
        if not isinstance(index, int):
            raise TypeError("index must be an integer")
        if index < 0 or index >= self._size:
            raise IndexError("index out of range")
        if index == 0:
            return self.pop_front()
        if index == self._size - 1:
            return self.pop_back()
        current = self._head
        for _ in range(index - 1):
            current = current.next
        value = current.next.value
        current.next = current.next.next
        self._size -= 1
        return value

    def size(self):
        return self._size

    def is_empty(self):
        return self._size == 0

    def clear(self):
        self._head = None
        self._tail = None
        self._size = 0

    def copy(self):
        clone = LinkedList()
        current = self._head
        while current is not None:
            clone.push_back(current.value)
            current = current.next
        return clone

    def __iter__(self):
        current = self._head
        while current is not None:
            yield current.value
            current = current.next

    def __len__(self):
        return self._size
